### DISCUSSION TITLE
sealed without override

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#5938] [Language Ideas]
Is it possible to ```sealed``` a method that is not a virtual method that is being overridden?

If not, can we make ```override``` optional and be inferred by ```sealed```?

Before:
```
protected sealed override void foo () { }
```

After:
```
protected sealed void foo () { }
```

**Comment by FaustVX** [comment]
Edit: It's not really relevant to the question, I've misread it.

Every method is `sealed` by default in C#
In Java, it's the opposite way.

**Reply by 333fred** [reply]
Not quite. Every method is _non-virtual_ by default, which isn't the same thing. I also don't think it's relevant to the question: Tony is asking to remove an `override` keyword, not anything about methods that are not virtual. `override` methods are not sealed by default.

**Reply by FaustVX** [reply]
What are the keys differences between `sealed` and _non-virtual_ methods ?
And yes, I misread his question.

**Reply by 333fred** [reply]
A `sealed` method is a virtual method that cannot be overridden further by derived types. It cannot be called with a `call` instruction, and calling it goes through the standard virtual dispatch mechanisms (though the JIT can dispense with that in some cases). Non-virtual methods are exactly what they sound like. They can be called with a `call` instruction (though the compiler almost always emits a `callvirt` anyway, for the `null` check on the receiver), and calling them does not go through virtual function lookup. There is no devirtualization for the JIT to do, because the function isn't virtual at all.

**Reply by FaustVX** [reply]
Ok, thanks.

**Reply by TonyValenti** [reply]
@333fred Am I correct in my understanding that the only time ```sealed``` is used, ```override``` would also be used?

**Reply by 333fred** [reply]
On methods, I believe that is correct. However, I'm not a huge fan of conflating the keywords in this case: `sealed` means only "this this cannot be further extended", and it has that same meaning when applied to methods or types. I think `override sealed` is perfectly fine for "I want to override a method from the base type and prevent further extending of this method".
